File: src/web/src/types/submission.ts
/**
 * Submission-related TypeScript definitions for the Molecular Data Management and CRO Integration Platform.
 * These types represent submissions to Contract Research Organizations (CROs), including submission details,
 * quotes, and status tracking for the frontend application.
 * @packageDocumentation
 */

import { Experiment } from './experiment';
import { User } from './user';
import { Result } from './result';

/**
 * Enumeration of possible submission status values.
 * Tracks the lifecycle of submissions to Contract Research Organizations.
 */
export enum SubmissionStatus {
  /** Initial state when submission is created but not yet sent to CRO */
  PENDING = 'pending',
  /** Submission was rejected by the CRO */
  REJECTED = 'rejected',
  /** CRO has provided a quote for the submission */
  QUOTE_PROVIDED = 'quote_provided',
  /** Quote was rejected by the pharma user */
  QUOTE_REJECTED = 'quote_rejected',
  /** Quote was approved by the pharma user and work can begin */
  APPROVED = 'approved',
  /** Work is currently being performed by the CRO */
  IN_PROGRESS = 'in_progress',
  /** All work has been completed by the CRO */
  COMPLETED = 'completed',
  /** Submission was cancelled before completion */
  CANCELLED = 'cancelled'
}

/**
 * Interface representing a detail for a submission.
 * Contains custom fields for specific experimental requirements.
 */
export interface SubmissionDetail {
  /** Unique identifier for the detail */
  id: string;
  /** ID of the submission this detail belongs to */
  submission_id: string;
  /** Name of the detail field */
  detail_name: string;
  /** Value of the detail field */
  detail_value: string;
}

/**
 * Interface for creating a new submission detail.
 * Used when adding custom fields to a submission.
 */
export interface SubmissionDetailCreate {
  /** Name of the detail field */
  detail_name: string;
  /** Value of the detail field */
  detail_value: string;
}

/**
 * Interface representing a submission to a CRO.
 * Contains the core data structure for experiment submissions.
 */
export interface Submission {
  /** Unique identifier for the submission */
  id: string;
  /** ID of the experiment this submission is for */
  experiment_id: string;
  /** ID of the CRO user assigned to this submission */
  cro_id: number;
  /** Current status of the submission */
  status: SubmissionStatus;
  /** Timestamp when the submission was created */
  submitted_at: string;
  /** Timestamp when the submission was last updated */
  updated_at: string;
  /** Price quoted by the CRO (if a quote has been provided) */
  price?: number;
  /** Estimated turnaround time in days */
  turnaround_days?: number;
  /** Additional notes about the submission */
  notes?: string;
  /** Related experiment data */
  experiment?: Experiment;
  /** CRO user information */
  cro?: User;
  /** Custom details for this submission */
  details?: SubmissionDetail[];
}

/**
 * Interface representing a submission with detailed information including results.
 * Extends the base Submission interface with result data.
 */
export interface SubmissionDetailed extends Submission {
  /** Results uploaded for this submission */
  results?: Result[];
}

/**
 * Interface for creating a new submission.
 * Contains the fields required to submit an experiment to a CRO.
 */
export interface SubmissionCreate {
  /** ID of the experiment to submit */
  experiment_id: string;
  /** ID of the CRO to assign the submission to */
  cro_id: number;
  /** Custom details for the submission */
  details?: SubmissionDetailCreate[];
  /** Additional notes about the submission */
  notes?: string;
}

/**
 * Interface for updating an existing submission.
 * Contains fields that can be modified after creation.
 */
export interface SubmissionUpdate {
  /** ID of the CRO to assign the submission to */
  cro_id?: number;
  /** New status for the submission */
  status?: SubmissionStatus;
  /** Updated custom details */
  details?: SubmissionDetailCreate[];
  /** Updated notes */
  notes?: string;
}

/**
 * Interface for filtering submissions by various criteria.
 * Used in search and list views.
 */
export interface SubmissionFilter {
  /** Filter by experiment ID */
  experiment_id?: string;
  /** Filter by CRO ID */
  cro_id?: number;
  /** Filter by submission status */
  status?: SubmissionStatus;
  /** Filter for submissions created after this date */
  submitted_after?: string;
  /** Filter for submissions created before this date */
  submitted_before?: string;
  /** Page number for pagination */
  page?: number;
  /** Page size for pagination */
  page_size?: number;
  /** Field to sort by */
  sort_by?: string;
  /** Whether to sort in descending order */
  sort_desc?: boolean;
}

/**
 * Interface for paginated list of submissions response.
 * Used in API responses.
 */
export interface SubmissionListResponse {
  /** Array of submission items */
  items: Submission[];
  /** Total number of submissions matching the query */
  total: number;
  /** Current page number */
  page: number;
  /** Number of items per page */
  size: number;
  /** Current filters applied */
  filters: SubmissionFilter;
}

/**
 * Interface for CRO to provide a quote for a submission.
 * Used by CRO users to respond to submissions with pricing.
 */
export interface QuoteProvide {
  /** Quoted price for the requested work */
  price: number;
  /** Estimated turnaround time in business days */
  turnaround_days: number;
  /** Additional notes about the quote */
  notes?: string;
}

/**
 * Interface for pharma user to respond to a quote.
 * Used to approve or reject quotes from CROs.
 */
export interface QuoteResponse {
  /** Whether the quote is approved */
  approved: boolean;
  /** Optional notes about the approval/rejection */
  notes?: string;
}

/**
 * Interface for updating the status of a submission.
 * Used for status changes during the submission lifecycle.
 */
export interface SubmissionStatusUpdate {
  /** New status for the submission */
  status: SubmissionStatus;
  /** Notes about the status change */
  notes?: string;
}

File: src/backend/app/schemas/submission.py
"""
Pydantic schema models for submission data validation, serialization, and API responses
in the Molecular Data Management and CRO Integration Platform.

These schemas represent the data structures for experiment submissions to Contract Research
Organizations (CROs), including submission details, quotes, and status tracking.
"""

from typing import Optional, List, Dict, Any  # standard library
from uuid import UUID  # standard library
from datetime import datetime  # standard library

from pydantic import BaseModel, Field, validator  # pydantic version 2.0+

from ..constants import SubmissionStatus  # Internal import for status enumeration
from .experiment import ExperimentResponse  # Internal import for experiment schema
from .user import UserResponse  # Internal import for user schema

# List of valid submission status values for validation
SUBMISSION_STATUS_VALUES = [status.name for status in SubmissionStatus]


@validator('status', pre=True)
def validate_status(cls, v):
    """
    Validates that a submission status is one of the allowed values.
    
    Args:
        cls: The class being validated
        v: The value to validate
        
    Returns:
        The validated status if valid
        
    Raises:
        ValueError: If the status is not a valid SubmissionStatus value
    """
    if v is None:
        return SubmissionStatus.PENDING.name
    
    try:
        # Convert to uppercase for case-insensitive comparison
        v_upper = v.upper() if isinstance(v, str) else v
        # Check if it's a valid enum value
        if v_upper not in SUBMISSION_STATUS_VALUES:
            raise ValueError(f"Invalid status: {v}. Must be one of {SUBMISSION_STATUS_VALUES}")
        return v_upper
    except (AttributeError, ValueError):
        raise ValueError(f"Invalid status: {v}. Must be one of {SUBMISSION_STATUS_VALUES}")


@validator('details', pre=True)
def validate_details(cls, v):
    """
    Validates that submission details are properly formatted.
    
    Args:
        cls: The class being validated
        v: The value to validate (list of detail dictionaries)
        
    Returns:
        The validated details list if valid
        
    Raises:
        ValueError: If details are not properly formatted
    """
    if v is None:
        return []
    
    if not isinstance(v, list):
        raise ValueError("Details must be a list")
    
    for detail in v:
        if not isinstance(detail, dict):
            raise ValueError("Each detail must be a dictionary")
        
        if 'detail_name' not in detail or not detail['detail_name']:
            raise ValueError("Each detail must have a non-empty 'detail_name'")
        
        if 'detail_value' not in detail:
            raise ValueError("Each detail must have a 'detail_value'")
    
    return v


class SubmissionDetailBase(BaseModel):
    """
    Base Pydantic model for submission detail data with common fields.
    
    This model defines the basic structure for submission details, including
    the detail name and value fields.
    """
    detail_name: str = Field(..., description="Name of the submission detail")
    detail_value: str = Field(..., description="Value of the submission detail")


class SubmissionDetailCreate(SubmissionDetailBase):
    """
    Pydantic model for creating a new submission detail.
    
    Extends the base model with a reference to the submission it belongs to.
    """
    submission_id: UUID = Field(..., description="ID of the submission this detail belongs to")


class SubmissionDetailRead(SubmissionDetailBase):
    """
    Pydantic model for submission detail data in API responses.
    
    Extends the base model with additional fields needed for API responses,
    such as the detail ID and submission ID.
    """
    id: UUID = Field(..., description="Unique identifier for the detail")
    submission_id: UUID = Field(..., description="ID of the submission this detail belongs to")


class SubmissionBase(BaseModel):
    """
    Base Pydantic model for submission data with common fields.
    
    This model defines the basic structure for submissions, including the
    experiment ID, CRO ID, status, and details fields.
    """
    experiment_id: UUID = Field(..., description="ID of the experiment being submitted")
    cro_id: Optional[UUID] = Field(None, description="ID of the CRO user assigned to this submission")
    status: Optional[str] = Field(SubmissionStatus.PENDING.name, 
                               description="Current status of the submission")
    details: Optional[List[SubmissionDetailCreate]] = Field(default_factory=list, 
                                                        description="List of submission details")
    notes: Optional[str] = Field(None, description="Additional notes or instructions")


class SubmissionCreate(SubmissionBase):
    """
    Pydantic model for creating a new submission.
    
    Inherits all fields from SubmissionBase.
    """
    pass


class SubmissionUpdate(BaseModel):
    """
    Pydantic model for updating an existing submission.
    
    All fields are optional to allow partial updates.
    """
    cro_id: Optional[UUID] = Field(None, description="Updated CRO user ID")
    status: Optional[str] = Field(None, description="Updated submission status")
    details: Optional[List[SubmissionDetailCreate]] = Field(None, 
                                                        description="Updated submission details")
    notes: Optional[str] = Field(None, description="Updated notes")


class SubmissionInDBBase(SubmissionBase):
    """
    Base Pydantic model for submission data as stored in the database.
    
    Extends the base model with additional fields that are present in the database,
    such as ID, timestamps, and pricing information.
    """
    id: UUID = Field(..., description="Unique identifier for the submission")
    submitted_at: datetime = Field(..., description="Timestamp when the submission was created")
    updated_at: Optional[datetime] = Field(None, 
                                        description="Timestamp when the submission was last updated")
    price: Optional[float] = Field(None, description="Price quoted by the CRO in USD")
    turnaround_days: Optional[int] = Field(None, description="Estimated turnaround time in days")


class SubmissionRead(SubmissionInDBBase):
    """
    Pydantic model for submission data in API responses.
    
    Extends the database model with additional fields needed for API responses,
    such as the experiment and CRO details.
    """
    experiment: ExperimentResponse = Field(..., description="Details of the experiment")
    cro: Optional[UserResponse] = Field(None, description="Details of the assigned CRO")
    details: List[SubmissionDetailRead] = Field(default_factory=list, 
                                             description="List of submission details")


class SubmissionDetailedRead(SubmissionInDBBase):
    """
    Pydantic model for detailed submission data in API responses.
    
    Extends the SubmissionRead model with additional fields for more detailed views,
    such as experiment results.
    """
    experiment: ExperimentResponse = Field(..., description="Details of the experiment")
    cro: Optional[UserResponse] = Field(None, description="Details of the assigned CRO")
    details: List[SubmissionDetailRead] = Field(default_factory=list, 
                                             description="List of submission details")
    results: Optional[List[Dict]] = Field(default_factory=list, 
                                       description="List of experimental results if available")


class SubmissionList(BaseModel):
    """
    Pydantic model for paginated list of submissions in API responses.
    
    Used for paginated responses containing multiple submissions.
    Includes the total count, page number, and page size for pagination.
    """
    items: List[SubmissionRead] = Field(..., description="List of submissions")
    total: int = Field(..., description="Total number of submissions matching the criteria")
    page: int = Field(..., description="Current page number")
    size: int = Field(..., description="Number of items per page")
    filters: Optional[Dict[str, Any]] = Field(None, description="Applied filters")


class SubmissionFilter(BaseModel):
    """
    Pydantic model for filtering submissions by various criteria.
    
    This model defines the available filter parameters that can be used
    when querying submissions. All fields are optional to allow flexible filtering.
    """
    experiment_id: Optional[UUID] = Field(None, description="Filter by experiment ID")
    cro_id: Optional[UUID] = Field(None, description="Filter by CRO user ID")
    status: Optional[str] = Field(None, description="Filter by submission status")
    submitted_after: Optional[datetime] = Field(None, 
                                             description="Filter by submission date (after)")
    submitted_before: Optional[datetime] = Field(None, 
                                              description="Filter by submission date (before)")
    skip: Optional[int] = Field(0, description="Number of submissions to skip (pagination)")
    limit: Optional[int] = Field(100, description="Maximum number of submissions to return")
    sort_by: Optional[str] = Field("submitted_at", 
                                 description="Field to sort by")
    sort_desc: Optional[bool] = Field(True, 
                                   description="Sort in descending order if true")
    
    @validator('status', pre=True)
    def validate_status(cls, v):
        """
        Validates that the status filter is one of the allowed values.
        
        Args:
            cls: The class being validated
            v: The value to validate
            
        Returns:
            The validated status if valid
            
        Raises:
            ValueError: If the status is not a valid SubmissionStatus value
        """
        if v is None:
            return None
        
        try:
            # Convert to uppercase for case-insensitive comparison
            v_upper = v.upper() if isinstance(v, str) else v
            # Check if it's a valid enum value
            if v_upper not in SUBMISSION_STATUS_VALUES:
                raise ValueError(f"Invalid status: {v}. Must be one of {SUBMISSION_STATUS_VALUES}")
            return v_upper
        except (AttributeError, ValueError):
            raise ValueError(f"Invalid status: {v}. Must be one of {SUBMISSION_STATUS_VALUES}")


class QuoteProvide(BaseModel):
    """
    Pydantic model for CRO to provide a quote for a submission.
    
    This model is used when a CRO responds to a submission request with pricing
    and estimated turnaround time.
    """
    price: float = Field(..., description="Quoted price in USD")
    turnaround_days: int = Field(..., description="Estimated turnaround time in days")
    notes: Optional[str] = Field(None, description="Additional notes or conditions")


class QuoteResponse(BaseModel):
    """
    Pydantic model for pharma user to respond to a quote.
    
    This model is used when a pharma user either approves or rejects a quote
    provided by a CRO.
    """
    approved: bool = Field(..., description="Whether the quote is approved")
    notes: Optional[str] = Field(None, description="Additional notes or explanation")


class SubmissionStatusUpdate(BaseModel):
    """
    Pydantic model for updating the status of a submission.
    
    This model is used to change the status of a submission (e.g., from
    IN_PROGRESS to COMPLETED).
    """
    status: str = Field(..., description="New submission status")
    notes: Optional[str] = Field(None, description="Additional notes about the status change")